qiime/core/type/signature.py

```diff
@@ -113,12 +114,20 @@ def decode_parameters(self, **kwargs):
         return params
 
     def check_types(self, **kwargs):
-        for key, (type_, _) in itertools.chain(self.inputs.items(),
-                                               self.parameters.items()):
+        for key, (type_, _) in self.inputs.items():
+            if kwargs[key] is None:
+                raise TypeError("Argument to input %r is None, but only "
+                                "parameters can have values of None." % key)
             if kwargs[key] not in type_:
                 raise TypeError("Argument to parameter %r is not a subtype of"
                                 " %r." % (key, type_))
 
+        for key, (type_, _) in self.parameters.items():
+            if kwargs[key] not in type_:
+                if not (self.defaults[key] is None and kwargs[key] is None):
```

Comment:
I think `self.defaults[key]` could raise a `KeyError` if the parameter doesn't have a default.